### PR TITLE
add dubbo-spring-boot-observability-starter

### PR DESCRIPTION
**Pull request by ShenFeng312**
## What is the purpose of the change
In https://github.com/apache/dubbo/pull/11021,Dubbo already supports Micrometer Observation API, But when it is not working with springboot3, you need to configure it yourself.
So I added dubbo-spring-boot-observability-starter to support springboot2.X or without `spring-boot-starter-actuator`
Details:
 ### Add default config for  `micrometer-observation` and `micrometer-tracing`
see `DubboTracingProperties`

### If the project is configured with `spring-boot-starter-actuator`, the configuration of `spring-boot-starter-actuator` is preferred

1. `BraveAutoConfiguration` is used to configure Brave when `spring-boot-starter-actuator` does not exist
2. `OpenTelemetryAutoConfiguration` is used to configure OpenTelemetry when `spring-boot-starter-actuator` does not exist
3. `DubboMicrometerTracingAutoConfiguration` is used to configure MicrometerTracing when `spring-boot-starter-actuator` does not exist
4. `DubboObservationAutoConfiguration` is used to configure Micrometer Observation when `spring-boot-starter-actuator` does not exist
5. `ObservationHandlerGrouping` is used to group `ObservationHandler` by type
6. `ObservationRegistryPostProcessor`  is used to group registry observationHandlers to observationConfig

I'm sorry for my bad english

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by ShenFeng312** [comment]
Hello! dubbo wants to add autoconfiguration of Micrometer Observation, I think you can help review this PR, can you help us @marcingrzejszczak

File: dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml [hunk 1]
```diff
@@ -140,9 +140,18 @@
         </dependency>
         <dependency>
             <groupId>io.micrometer</groupId>
-            <artifactId>micrometer-tracing-bridge-otel</artifactId>
+            <artifactId>micrometer-tracing-bridge-brave</artifactId>
```
Comment (marcingrzejszczak) [inline dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml:143]: Out of curiosity - why the change?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/annotation/ConditionalOnDubboTracingEnable.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * @author shenfeng
+ */
+@Target({ElementType.TYPE,ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+@ConditionalOnProperty(prefix = "dubbo.tracing", name = "enabled", matchIfMissing = true)
+public @interface ConditionalOnDubboTracingEnable {
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/annotation/ConditionalOnDubboTracingEnable.java:37]: `Enable` or `Enabled` ?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboMicrometerTracingAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import org.apache.dubbo.spring.boot.observability.annotation.ConditionalOnDubboTracingEnable;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.core.annotation.Order;
+
+/**
+ * copy form {@link org.springframework.boot.actuate.autoconfigure.tracing.MicrometerTracingAutoConfiguration}
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboMicrometerTracingAutoConfiguration.java:33]: You can mention that this class is available from Boot 3.0

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+import org.springframework.core.Ordered;
+import org.springframework.core.annotation.Order;
+import org.springframework.core.env.Environment;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * @author shenfeng
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java:68]: You can mention that this is a copy from Boot 3.0 setup and I would schedule this for removal once you upgrade to Boot 3 since you don't want to maintain this copy forever

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        return CurrentSpanCustomizer.create(tracing);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(io.micrometer.tracing.SpanCustomizer.class)
+    BraveSpanCustomizer braveSpanCustomizer(SpanCustomizer spanCustomizer) {
+        return new BraveSpanCustomizer(spanCustomizer);
+    }
+
+//    @Configuration(proxyBeanMethods = false)
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java:157]: Why is this commented out?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@AutoConfiguration(before = DubboMicrometerTracingAutoConfiguration.class, afterName = "org.springframework.boot.actuate.autoconfigure.tracing.OpenTelemetryAutoConfiguration")
+@ConditionalOnDubboTracingEnable
+@ConditionalOnClass({OtelTracer.class, SdkTracerProvider.class, OpenTelemetry.class})
+@EnableConfigurationProperties(DubboTracingProperties.class)
+public class OpenTelemetryAutoConfiguration {
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:72]: Same remarks as for Brave

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author shenfeng
+ */
+@ConfigurationProperties(prefix = "dubbo.tracing")
+public class DubboTracingProperties {
+    private boolean enable = true;
+    private String baggage = "W3C";
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java:31]: is this even used?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author shenfeng
+ */
+@ConfigurationProperties(prefix = "dubbo.tracing")
+public class DubboTracingProperties {
+    private boolean enable = true;
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java:30]: `enabled` instead of `enable` ?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  "groups": [
+    {
+      "name": "dubbo.tracing",
+      "type": "org.apache.dubbo.spring.boot.observability.config.DubboTracingProperties",
+      "sourceType": "org.apache.dubbo.spring.boot.observability.config.DubboTracingProperties"
+    }
+  ],
+  "properties": [
+    {
+      "name": "dubbo.tracing.baggage",
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json:11]: Is this related to baggage really or context propagation type? Cause how baggage is done is related to context propagation type really

File: dubbo-spring-boot/pom.xml [hunk 10]
```diff
@@ -57,6 +58,11 @@
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>
+            <dependency>
+                <groupId>io.micrometer</groupId>
+                <artifactId>micrometer-core</artifactId>
+                <version>1.10.3</version>
```
Comment (marcingrzejszczak) [inline dubbo-spring-boot/pom.xml:64]: 1.10.4 is the latest

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/annotation/ConditionalOnDubboTracingEnable.java (hunk 2 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/annotation/ConditionalOnDubboTracingEnable.java:37]: Enabled,I'm sorry I forgot to test this feature

File: dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml (hunk 1 above)
Comment (ShenFeng312) [reply dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml:143]: I tested on the demo, but I forgot to rollback

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboMicrometerTracingAutoConfiguration.java (hunk 3 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboMicrometerTracingAutoConfiguration.java:33]: thank you

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java (hunk 8 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java:30]: yes!  thank you ,I'm sorry for my bad english

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java (hunk 5 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/brave/BraveAutoConfiguration.java:157]: > Why is this commented out?

I forget to clean this,I'm so sorry

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java (hunk 7 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java:31]: it is used in  `BraveAutoConfiguration` and `OpenTelemetryAutoConfiguration`

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json (hunk 9 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json:11]: can i rename it to dubbo.tracing.propagation?

Comment (marcingrzejszczak) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json]: This is how we do it in Boot - https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java#L171 `management.tracing.propagation.type`

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java (hunk 7 above)
Comment (marcingrzejszczak) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/config/DubboTracingProperties.java]: AFAIR that code was commented out

File: dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml (hunk 1 above)
Comment (marcingrzejszczak) [reply dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-consumer/pom.xml]: No problem, it doesn't matter which one you're using

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json (hunk 9 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/resources/META-INF/spring-configuration-metadata.json]: > This is how we do it in Boot - https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/TracingProperties.java#L171 `management.tracing.propagation.type`

OK

File: dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-provider/pom.xml [hunk 11]
```diff
@@ -141,7 +141,15 @@
             <groupId>io.micrometer</groupId>
             <artifactId>micrometer-tracing-bridge-otel</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.apache.dubbo</groupId>
+            <artifactId>dubbo-spring-boot-observability-starter</artifactId>
+        </dependency>
 
+<!--        <dependency>-->
```
[the remark this answers is not in the document]
Comment (ShenFeng312) [reply dubbo-demo/dubbo-demo-spring-boot/dubbo-demo-spring-boot-provider/pom.xml]: yes

**Review by AlbumenJ — APPROVED** [review]
LGTM
@songxiaosheng @conghuhu PTAL

File: dubbo-spring-boot/pom.xml [hunk 12]
```diff
@@ -57,6 +58,11 @@
                 <type>pom</type>
                 <scope>import</scope>
             </dependency>
+            <dependency>
+                <groupId>io.micrometer</groupId>
+                <artifactId>micrometer-core</artifactId>
+                <version>1.10.4</version>
```
Comment (conghuhu) [inline dubbo-spring-boot/pom.xml:64]: Should put the version in<properties>?

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboObservationAutoConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Register observationRegistry to ApplicationModel.
+ * Create observationRegistry when you are using Boot <3.0 or you are not using spring-boot-starter-actuator
+ * @author shenfeng
+ */
+@AutoConfiguration(after =DubboMicrometerTracingAutoConfiguration.class,afterName = "org.springframework.boot.actuate.autoconfigure.observation.ObservationAutoConfiguration")
+@ConditionalOnDubboTracingEnable
+public class DubboObservationAutoConfiguration implements BeanFactoryAware, SmartInitializingSingleton {
+
+    public DubboObservationAutoConfiguration( ApplicationModel applicationModel){
```
Comment (conghuhu) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/DubboObservationAutoConfiguration.java:55]: Remove space.

**Review by songxiaosheng — CHANGES_REQUESTED** [review]

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+    @Bean
+    @ConditionalOnMissingBean
+    Sampler otelSampler() {
+        Sampler rootSampler = Sampler.traceIdRatioBased(this.dubboTracingProperties.getSampling().getProbability());
+        return Sampler.parentBased(rootSampler);
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    SpanProcessor otelSpanProcessor(ObjectProvider<SpanExporter> spanExporters,
```
Comment (conghuhu) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:123]: I see that there is no @ConditionalOnMissingBean in the latest springboot3: 
image

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+                                    ObjectProvider<SpanFilter> spanFilters) {
+        return BatchSpanProcessor.builder(new CompositeSpanExporter(spanExporters.orderedStream().collect(Collectors.toList()),
+            spanExportingPredicates.orderedStream().collect(Collectors.toList()), spanReporters.orderedStream().collect(Collectors.toList()),
+            spanFilters.orderedStream().collect(Collectors.toList()))).build();
+    }
+
+    @Bean
+    @ConditionalOnMissingBean
+    Tracer otelTracer(OpenTelemetry openTelemetry) {
+        return openTelemetry.getTracer("org.apache.dubbo");
```
Comment (conghuhu) [inline dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:134]: Could we add the Dubbo version:
image

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java (hunk 14 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:123]: `OpenTelemetryAutoConfiguration` is after `org.springframework.boot.actuate.autoconfigure.tracing.OpenTelemetryAutoConfiguration`,so we need add  `@ConditionalOnMissingBean` use springboot3 default

Comment (conghuhu) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:123]: Okay, I see.

File: dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java (hunk 15 above)
Comment (ShenFeng312) [reply dubbo-spring-boot/dubbo-spring-boot-observability-starter/src/main/java/org/apache/dubbo/spring/boot/observability/autoconfigure/otel/OpenTelemetryAutoConfiguration.java:134]: Yes

**Comment by ShenFeng312** [comment]
Shoud I add dubbo-metrics-default in dubbo-spring-boot-observability-starter?  @conghuhu @AlbumenJ @songxiaosheng

**Comment by songxiaosheng** [comment]
> Shoud I add dubbo-metrics-default in dubbo-spring-boot-observability-starter?  @conghuhu @AlbumenJ @songxiaosheng 

add

**Comment by songxiaosheng** [comment]
> Shoud I add dubbo-metrics-default in dubbo-spring-boot-observability-starter?  @conghuhu @AlbumenJ @songxiaosheng 

We need to consider our options actuator dependency or   the observer starter uses an observability dependency

**Review by conghuhu — COMMENTED** [review]
You should remove all the author information

**Review by conghuhu — APPROVED** [review]
LGTM  @songxiaosheng

**Review by songxiaosheng — APPROVED** [review]
LGTM
